src/evaluation/evaluator.h

```diff
@@ -1,312 +1,271 @@
 #pragma once
 
 #include "engine/move_handling.h"
+#include "engine/thread_pool.h"
 #include "engine/tt_hash_table.h"
 #include "evaluation/move_ordering.h"
 #include "evaluation/pv_table.h"
 #include "evaluation/repetition.h"
 #include "evaluation/static_evaluation.h"
+
 #include "fmt/ranges.h"
 #include "movegen/move_types.h"
 #include "syzygy/syzygy.h"
 #include <atomic>
 #include <engine/zobrist_hashing.h>
 
 #include <chrono>
+#include <future>
 #include <mutex>
-#include <thread>
 
 namespace evaluation {
 
-class Evaluator {
-public:
-    constexpr movegen::Move getBestMove(const BitBoard& board, std::optional<uint8_t> depthInput = std::nullopt)
-    {
-        m_startTime = std::chrono::system_clock::now();
+enum SearchType {
+    Default,
+    NullSearch,
+};
 
-        /*
-         * if a depth has been provided then make sure that we search to that depth
-         * timeout will be 10 minutes
-         */
-        if (depthInput.has_value()) {
-            m_endTime = m_startTime + std::chrono::minutes(10);
-        } else {
-            setupTimeControls(m_startTime, board);
-        }
+struct MoveResult {
+    uint64_t hash;
+    BitBoard board;
+};
 
-        m_hash = engine::generateHashKey(board);
+struct SearcherResult {
+    int32_t score;
+    uint8_t searcherId;
+    movegen::Move pvMove;
+    uint8_t searchedDepth;
+};
 
-        /* we're dependent on the hash table - ensure it's initialized! */
-        if (engine::TtHashTable::getSizeMb() == 0) {
-            engine::TtHashTable::setSizeMb(s_defaultTtHashTableSizeMb);
-        }
+template<uint8_t maxSize>
+class MoveVoteMap {
 
-        return scanForBestMove(depthInput.value_or(s_maxSearchDepth), board);
+    using MoveVotePair = std::pair<movegen::Move, uint8_t>;
+
+public:
+    MoveVoteMap() = default;
+
+    MoveVoteMap(std::initializer_list<MoveVotePair> init)
+    {
+        assert(init.size() < maxSize);
+
+        for (auto& entry : init) {
+            m_entries.at(m_size++) = std::move(entry);
+        }
     }
 
-    constexpr std::optional<movegen::Move> getPonderMove() const
+    constexpr uint8_t& at(const movegen::Move& ext_move)
```

Comment:
If use for lookup only, let's add a const non-reference variant as well

Reply:
The only use of `at` is in the voting, and we want the value (counter) to be mutable. So I think the signature is correct here?